### PR TITLE
8291598: Matcher.appendReplacement should not create new StringBuilder instances

### PR DESCRIPTION
**Pull request by rgiulietti**
Remove instantiation of `StringBuilder`

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8291598](https://bugs.openjdk.org/browse/JDK-8291598): Matcher.appendReplacement should not create new StringBuilder instances

### Reviewers
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13048/head:pull/13048` \
`$ git checkout pull/13048`

Update a local copy of the PR: \
`$ git checkout pull/13048` \
`$ git pull https://git.openjdk.org/jdk.git pull/13048/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13048`

View PR using the GUI difftool: \
`$ git pr show -t 13048`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13048.diff">https://git.openjdk.org/jdk/pull/13048.diff</a>

</details>

### CODE REVIEW

**Comment by rgiulietti** [comment]
Benchmark for the example in the documentation for `Matcher.appendReplacement()`

before
```
Benchmark                                Mode  Cnt    Score   Error  Units
AppendReplacement.testAppendReplacement  avgt   15  177.029 ± 6.294  ns/op
```
after
```
Benchmark                                Mode  Cnt    Score   Error  Units
AppendReplacement.testAppendReplacement  avgt   15  142.373 ± 1.684  ns/op
```

Same example, but with pattern `"(cat)"` (1 capturing group), matcher on the input repeated 1000 times, and replacement string `"$1dog$1cat$1mouse"` (3 back references).

before
```
Benchmark                                Mode  Cnt       Score      Error  Units
AppendReplacement.testAppendReplacement  avgt   15  262576.335 ± 3718.905  ns/op
```
after
```
Benchmark                                Mode  Cnt       Score     Error  Units
AppendReplacement.testAppendReplacement  avgt   15  225700.642 ± 683.721  ns/op
```

File: src/java.base/share/classes/java/util/regex/Matcher.java [hunk 1]
```diff
@@ -1021,93 +1029,95 @@ public Matcher appendReplacement(StringBuilder sb, String replacement) {
...
+                        while (cursor < replacement.length()) {
+                            nextChar = replacement.charAt(cursor);
+                            if (ASCII.isLower(nextChar) ||
+                                    ASCII.isUpper(nextChar) ||
+                                    ASCII.isDigit(nextChar)) {
+                                gsb.append(nextChar);
+                                cursor++;
+                            } else {
+                                break;
+                            }
```
[the remark this answers is not in the document]
Comment (rgiulietti) [reply src/java.base/share/classes/java/util/regex/Matcher.java]: Thanks @RogerRiggs, addressed in latest commit.

**Comment by rgiulietti** [comment]
Waiting for other comments and review

**Review by RogerRiggs — APPROVED** [review]
Sorry for the delay.
The workaround for appending to StringBuffer/StringBuilder is awkward, adding a try/catch that can't happen.  But I don't have a better suggestion. 
It is scoped to the appendExpandedReplacement method and has no runtime impact.

**Comment by rgiulietti** [comment]
@RogerRiggs Yes, it would be nice to be able to write
```
private void appendExpandedReplacement(StringBuilder | StringBuffer app, String replacement) 
```
and get rid of the `try` statement altogether.

**Pull request closed without merging**